### DISCUSSION TITLE
Continuous Delivery core app versioning proposal (for initial App Hub upload)

### DISCUSSION THREAD

**Discussion by amcgee** [dhis2/notes#268] [Specs & RFCs]
## Scope

This is not intended to address the larger issue of versioning, release management, QA coordination, and Jira configuration for independently-versioned applications on the DHIS2 app hub.  Rather, it is intended to propose an "MVP" strategy we can use to upload all core applications to the app hub, the goal being primarily to create an App Hub ID for each Core application.

## Proposal

As the goal here is to create an App Hub ID for each core app, the proposal is quite simple:

1. Create a new App Hub app for the core application, uploading the latest version from the master branch with the parameters `organization: DHIS2`, `name: <name>`, `version: 0.0.0`, `channel: dev`
2. Set the `id` parameter in `d2.config.js` to the generated app id
3. Run `d2-app-scripts publish --version 0.1.0 --channel dev` to upload a new version with the correct id specified

Min and max DHIS2 versions should also be injected into d2.config.js 

As an alternative, we could simply
1. Remove the requirement for uploading a version when creating an app
2. Create an app for each core app and update the id in `d2.config.js`
3. Don't upload any version

In both of these alternatives, publication of new versions to the app hub app should be deferred until a versioning and release process scheme has been ratified (see next steps)

## Next steps

We need to figure out the full release lifecycle and versioning workflow for these apps, see [this document](https://docs.google.com/document/d/1LXi2-CeIAtZwTntFcVXC1EbaUT1oLL3CEqvqcqmAr_s/edit) which needs to be refreshed and turned into an actual proposal.

Particular consideration needs to be given to apps which DO currently have meaningful version numbers (like `data-visualizer`) and/or can't yet adopt feature toggling.

**Comment by varl** [comment]
Sorry I missed the meeting today.

I have looked into this and the best option I have found is to _not_ upload a version until we have settled on a versioning strategy. Any other movement implies a direction that will be hard to back out of after the fact.

For the proposal, step 1 & 2 can be done without App Hub changes. We should omit step 3 for now.

`0.0.0` is considered a "not-a-version". The first uploaded version must have a positive integer in the version string.

The alternative can be done as is, but requires App Hub changes.

**Reply by Philip-Larsen-Donnelly** [reply]
I think we concluded the same in the meeting - i.e. that we should not upload until we have settled on the versioning strategy.
